src/garage/torch/algos/dqn.py

```diff
@@ -0,0 +1,268 @@
+                               np.mean(self._episode_qf_losses))
+                tabular.record('QFunction/AverageQ', np.mean(self._epoch_qs))
+                tabular.record('QFunction/MaxQ', np.max(self._epoch_qs))
+                tabular.record('QFunction/AverageAbsQ',
+                               np.mean(np.abs(self._epoch_qs)))
+                tabular.record('QFunction/AverageY', np.mean(self._epoch_ys))
+                tabular.record('QFunction/MaxY', np.max(self._epoch_ys))
+                tabular.record('QFunction/AverageAbsY',
+                               np.mean(np.abs(self._epoch_ys)))
+
+        if itr % self._target_update_freq == 0:
+            self._target_qf = copy.deepcopy(self._qf)
+
+    def optimize_qf(self, samples_data):
+        """Perform algorithm optimizing.
+
+        Args:
+            samples_data (dict): Processed batch data.
+
+        Returns:
+            qval_loss: Loss of Q-value predicted by the Q-network.
+            ys: y_s.
+            qval: Q-value predicted by the Q-network.
+
+        """
+        transitions = dict_np_to_torch(samples_data)
+
+        observations = transitions['observations']
+        rewards = transitions['rewards'].reshape(-1, 1)
+        actions = transitions['actions']
+        next_observations = transitions['next_observations']
+        terminals = transitions['terminals'].reshape(-1, 1)
+
+        next_inputs = next_observations
+        inputs = observations
+        with torch.no_grad():
+            # discrete, outputs Qs for all possible actions
+            target_qvals = self._target_qf(next_inputs)
+            best_qvals, _ = torch.max(target_qvals, 1)
+            best_qvals = best_qvals.unsqueeze(1)
+
+        rewards_clipped = rewards
+        if self._clip_reward is not None:
+            rewards_clipped = torch.clamp(rewards, -1 * self._clip_reward,
+                                          self._clip_reward)
+        y_target = (rewards_clipped +
+                    (1.0 - terminals) * self._discount * best_qvals)
+        y_target = y_target.squeeze(1)
+
+        # optimize qf
+        qvals = self._qf(inputs)
+        selected_qs = torch.sum(qvals * actions, axis=1)
+        qval_loss = F.smooth_l1_loss(selected_qs, y_target)
+
+        self._qf_optimizer.zero_grad()
+        qval_loss.backward()
+
+        # optionally clip the gradients
+        if self._clip_grad is not None:
+            torch.nn.utils.clip_grad_norm_(self.policy.parameters(),
+                                           self._clip_grad)
+        self._qf_optimizer.step()
+
+        return (qval_loss.detach(), y_target, selected_qs.detach())
+
+    def to(self, device=None):
+        """Put all the networks within the model on device.
+
+        Args:
+            device (str): ID of GPU or CPU.
+
+        """
+        if device is None:
+            device = global_device()
+        torch.set_default_tensor_type(torch.cuda.FloatTensor)
```

Comment:
does this work on CPU only devices? Can you confirm this? It makes more sense to use our set_gpu function than this line over here.

Reply:
I needed both `set_gpu` (which is called in the launcher) and setting the default tensor type. Confirmed it works on CPU only devices.